### ISSUE TITLE
Add support for custom search query

### ISSUE THREAD

**Issue by theo-gk** [superspeedyplugins/issue-tracker#16]
**Is your feature request related to a problem? Please describe.**
A client decided to offload the search functionality to a 3rd party service. After the search form is submitted, the user gets redirected to the search page (as usual) with some extra param to the URL like this:  **mysite.gr/?s=adidas&category=mens+sneakers**
So I send to their API the search terms ('s' term, category) from the URL and their API sends me back the WP Post IDs.

**If I use native WC filters, I can make it work** by doing both (1) and (2):

1. Customize main search query by doing both (a) and (b):
a)  Using `pre_get_posts` filter, and unsetting ['s'] query var so I can handle search myself in all cases, and for example WC_Query::get_main_search_query_sql() will be empty
b) Using `pre_get_posts` filter and setting a ['post__in'] var, so I can show only the products from the API results

2. Modify WC filters to respect my custom results by:
a) Modify filter counts SQL query with filter `woocommerce_get_filtered_term_product_counts_query`
Code: 
```php
// modify filter counts to respect custom search
add_filter( 'woocommerce_get_filtered_term_product_counts_query', 'dc_modify_filters_count_query' );
function dc_modify_filters_count_query( $query ) {
	
	global $wpdb;

	// calculate IDs here
	$custom_search_ids = [ 350990, 115451 ];
	
	$custom_search_ids_sql = "{$wpdb->posts}.ID IN (" . implode( ',', $custom_search_ids ) . ")";
	
	$query['where'] .= ' AND ' . $custom_search_ids_sql;
	
	return $query;
}
``` 

b) Modify filter current page URL with filter `woocommerce_widget_get_current_page_url` -> Code: https://prnt.sc/5xeW50rjyx6X
Code: 
```php
add_filter( 'woocommerce_widget_get_current_page_url', 'dc_add_custom_params_to_filters_base_url' );
function dc_add_custom_params_to_filters_base_url( $link ) {
	
	if ( isset( $_GET['category'] ) ) {
		$link = add_query_arg( 'category', wc_clean( $_GET['category'] ), $link );
	}
	
	if ( isset( $_GET['manufacturer'] ) ) {
		$link = add_query_arg( 'manufacturer', wc_clean( $_GET['manufacturer'] ), $link );
	}
	
	return $link;
}
``` 

**Describe the solution you'd like**
With SSF enabled (no AJAX), I can't do customizations in (2) anymore. 
As a result, I have 2 problems:
1. The counts are wrong (the post__in clause is not taken into account)
2. If I click on a filter option, then the custom param is lost from the URL, because the filter base URL does not contain my custom parameters

The customizations from (1) work with SSF, so the search results are fine. 
Also the filtering actually works, so I can filter for "adidas" correctly, but the counts are wrong as I mentioned.

It would be nice if there was a filter like in WC where I can modify the SQL query.
It would be nice if there was a filter like in WC where I add my custom parameters to the filters base URL, if they exist in the URL.

Thank you!

**Comment by dhilditch** [comment]
I'll look at this myself:

1. Figure out how to achieve 1 & 2 from above
2. Create KB guide showing how to customise the queries

**Comment by dhilditch** [comment]
@theo-gk can you please re-test your code and use a priority for your filter higher than 10? My developer reports that this is already working and we just need to update the docs - can you confirm please?

**Comment by theo-gk** [comment]
I checked again with priority 9999.
Search results are correct. Maybe that's the part your dev said that is working? 

I have problem with the counts and the URL:
> As a result, I have 2 problems:
> 1. The counts are wrong (the post__in clause is not taken into account)
> 2. If I click on a filter option, then the custom param is lost from the URL, because the filter base URL does not contain my custom > parameters

I can't see these 2 working right now :/ @dhilditch

**Comment by voodoo911** [comment]
@theo-gk 
Hello, can you pls provide example of code you use to replicate 'pre_get_posts' issue you have ?
Tnx

**Comment by theo-gk** [comment]
> @theo-gk Hello, can you pls provide example of code you use to replicate 'pre_get_posts' issue you have ? Tnx

Sorry for late reply. My pre_get_posts query is very simple. The API I use for search results returns some product IDs, so I just insert them to the search query.

```
// modify query results
add_action( 'pre_get_posts', 'dc_example_query', 9999 );
function dc_example_query( $query ) {

    $qv = &$query->query_vars;
    
    if ( ! $query->is_main_query() || ! $query->is_search ) return;
    
    // unset original search param
    $qv['s'] = ''; 
    $query->query['s'] = '';
    
    // add custom post__in var to limit search to my custom results
    // calc custom IDS to show - THIS IS AN EXAMPLE WITH 2 IDs
    $custom_search_ids = [ 350990, 115451 ];
    $query->set('post__in', $custom_search_ids);
}
```
This works fine as you can see here, but the counts don't: [https://prnt.sc/Nz5FF5VsSqRr](https://prnt.sc/Nz5FF5VsSqRr)

To make the counts work, for native WC filter widgets I use this:
```
// modify filter counts to respect custom search
add_filter( 'woocommerce_get_filtered_term_product_counts_query', 'dc_modify_filters_count_query', 9999 );
function dc_modify_filters_count_query( $query ) {
	
	global $wpdb;

        // THIS IS AN EXAMPLE WITH 2 IDs
	$custom_search_ids = [ 350990, 115451 ];
	$custom_search_ids_sql = "{$wpdb->posts}.ID IN (" . implode( ',', $custom_search_ids ) . ")";
	
	$query['where'] .= ' AND ' . $custom_search_ids_sql;
	
	return $query;
}
```
But this of course doesn't work for your widgets, because it's for WooCommerce only. I need something similar, or some other code from you guys to make counts work.

***
Problem 2 as I described on original post: My custom URL params are lost from the filtered links. I need to keep them
[https://prnt.sc/WTELDAQgGq22](https://prnt.sc/WTELDAQgGq22)

Again, in WooCoomerce I can do the following, which doesn't work in your plugin:
```
// Add custom params to filter URLs
add_filter( 'woocommerce_widget_get_current_page_url', 'dc_add_custom_params_to_filters_base_url', 9999 );
function dc_add_custom_params_to_filters_base_url( $link ) {

        // my own custom Param
	if ( isset( $_GET['manufacturer'] ) ) {
		$link = add_query_arg( 'manufacturer', wc_clean( $_GET['manufacturer'] ), $link );
	}
	
	return $link;
}
```

Thank you, I hope it's more clear now.
@voodoo911

**Comment by theo-gk** [comment]
UPDATE:
Just saw a message from 15 Feb by @dhilditch in discord about a new filter you added. I'm sorry if you posted this before but I missed it, but I was not aware that this filter was added.

When I opened the ticket, the version was 5.14 which did not contain the filter. 
I just updated to 5.19 and I added the code like this:
```
add_filter('ssf_widget_query_filter', function() {
        // example for my 2 custom IDs
	return ' AND p.ID in (350990,115451) '; 
});
```
And yes, the counts now work [https://prnt.sc/hbgjZa5JB8Tr](https://prnt.sc/hbgjZa5JB8Tr)

***
I still need help with the URLs (Problem 2 from previous comment)
@voodoo911

**Comment by voodoo911** [comment]
@theo-gk 
Hello, in version 5.19 you can use filter ssf_custom_url_parameters and pass there attributes as array.
Here is piece of code from plugin

$custom_parameters = apply_filters('ssf_custom_url_parameters', []);
foreach( $custom_parameters as $s_parameter ){
         $newurl = add_query_arg( $s_parameter, $this->linkfilters[ $s_parameter ],  $newurl);
}

**Comment by theo-gk** [comment]
Hello, this filter doesn't exist in version 5.19. It doesn't appear anywhere in plugin files when I serach with PHPStorm, and also the code doesn't work. 

Maybe you mean a version after 5.19? In my dashboard I see up to 5.19 though....
Could you please check again?
Thanks @voodoo911

**Comment by voodoo911** [comment]
@theo-gk Seems that fix was not merged yet

**Comment by dhilditch** [comment]
Fixed in 5.20, merged, code should be deployed tomorrow once system testing fully complete.
